src/keyring.rs

```diff
@@ -22,26 +53,43 @@ impl Keyring {
         self.keys.is_empty()
     }
 
-    pub fn keys(&self) -> &[Key] {
-        &self.keys
+    /// A vector with reference to all the keys in the keyring.
+    ///
+    /// All current uses only need references to keys, so provide this directly.
+    pub fn keys(&self) -> Vec<&T> {
+        self.keys.iter().collect()
```

Comment:
BLOCKING: this shouldn't allocate, but still return a slice

Reply:
Oh, fair enough.  Turns out the comment is an over-generalisation and I only needed to allocate in one place for this.